swarm/src/handler/either.rs

```diff
@@ -178,102 +176,144 @@ where
                 protocol: output,
                 info,
             }) => match (self, output, info) {
-                (Either::Left(handler), EitherOutput::First(output), Either::Left(info)) =>
-                {
-                    #[allow(deprecated)]
-                    handler.inject_fully_negotiated_outbound(output, info)
+                (Either::Left(handler), EitherOutput::First(output), Either::Left(info)) => {
+                    handler.on_connection_event(ConnectionEvent::FullyNegotiatedOutbound(
```

Comment:
yeah, that would be great. I tried:
```
impl<'a, R, L>
    ConnectionEvent<
        'a,
        EitherUpgrade<SendWrapper<L::InboundProtocol>, SendWrapper<R::InboundProtocol>>,
        EitherUpgrade<SendWrapper<L::OutboundProtocol>, SendWrapper<R::OutboundProtocol>>,
        Either<L::InboundOpenInfo, R::InboundOpenInfo>,
        Either<L::OutboundOpenInfo, R::OutboundOpenInfo>,
    >
where
    L: ConnectionHandler,
    R: ConnectionHandler,
{}
```
but get that `L` and `R` are not constrained.  I wonder if https://github.com/libp2p/rust-libp2p/pull/3085#discussion_r1016988595 would have changed that.

Reply:
I was thinking something like https://github.com/libp2p/rust-libp2p/commit/e3480d2db8916e7c9ba0115126880ea11d699f71. It doesn't compile yet but maybe you can make it work!

Reply:
thanks for the lead Thomas, managed to make it work. Used `Either::Left` for `AddressChange` instead of the `Result` as it seemed more semantically accurate in the sense that it's not err, it just doesn't matter if it's either `Left` or `Right` for that variant, cause when matching on `on_connection_event` we still have the `Either::Right` . Ptal and see if you agree.

Reply:
> thanks for the lead Thomas, managed to make it work. Used `Either::Left` for `AddressChange` instead of the `Result` as it seemed more semantically accurate in the sense that it's not err, it just doesn't matter if it's either `Left` or `Right` for that variant, cause when matching on `on_connection_event` we still have the `Either::Right` . Ptal and see if you agree.

Yeah I wasn't sure about that one. I think one could argue it is an error because we couldn't unambiguously transpose it. Can we make it work with `Result` or is moving to `Either::Left` essential?

Alternatively, we could wrap it in another `Either` but that is a bit weird. If possible, I'd prefer the `Result` return type and some documentation on why it is necessary.

Reply:
no you are right, documenting it informing that an `Err` can be because `AddressChange` isn't transposable makes sense to me, updated it :)